File: SUBARRAY-PRODUCT/SP.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        arrmax = max(nums)
        currmax = 1
        currmin = 1

        for i in nums:
            temp = currmax * i
            currmax = max({temp, currmin * i, i})
            currmin = min({temp, currmin * i, i})

            arrmax = max(currmax, arrmax)
        return arrmax

File: SUBARRAY-PRODUCT/SP.java
class Solution {
    public int maxProduct(int[] nums) {
        int arrmax = Arrays.stream(nums).max().getAsInt();
        int currmax = 1;
        int currmin = 1;

        for (int i : nums){
            int temp = currmax * i;
            currmax = Math.max(temp, Math.max(currmin * i, i));
            currmin = Math.min(temp, Math.min(currmin * i, i));

            arrmax = Math.max(arrmax, currmax);

        }
        return arrmax;
    }
}